File: go/service/proposal_summary.go
package service

import (
	"context"
	"crypto/sha256"
	"errors"
	"fmt"
	"io/fs"
	"sort"

	"github.com/rcarmo/memento/go/control"
	"github.com/rcarmo/memento/go/repository"
)

const maxProposalSummaryChanges = 200
const maxProposalSummaryAssets = 200
const maxProposalSummaryText = 2000

// Summary exposes proposal metadata without building a preview. The caller must
// authorise visibility before calling it; this is also used by review responses.
func (q ProposalQueue) Summary(ctx context.Context, record control.ProposalRecord, includeConflicts bool) (map[string]any, error) {
	assets, err := q.Proposals.ListAssets(ctx, control.ProposalAssetQuery{ProposalID: &record.ProposalID})
	if err != nil {
		return nil, err
	}
	return q.summary(ctx, record, assets, includeConflicts, defaultProposalRepository())
}
func (q ProposalQueue) summary(ctx context.Context, record control.ProposalRecord, allAssets []control.ProposalAssetRecord, includeConflicts bool, repo proposalRepository) (map[string]any, error) {
	patch, err := record.Patch()
	if err != nil {
		return nil, err
	}
	changes := []any{}
	if value, exists := patch["changes"]; exists {
		var ok bool
		changes, ok = value.([]any)
		if !ok {
			return nil, &ChangeValidationError{"proposal changes must be a list"}
		}
	}
	summaries := []any{}
	for i, value := range changes[:min(len(changes), maxProposalSummaryChanges)] {
		raw, ok := value.(map[string]any)
		if !ok {
			continue
		}
		item := map[string]any{"index": i, "kind": raw["kind"], "path": raw["path"]}
		if raw["kind"] == "rename" {
			item["new_path"] = raw["new_path"]
		}
		summaries = append(summaries, item)
	}
	assets := []any{}
	for _, asset := range allAssets[:min(len(allAssets), maxProposalSummaryAssets)] {
		manifest, err := asset.Manifest()
		if err != nil {
			return nil, err
		}
		matching := []string{}
		var bodyMatches any
		body, err := q.resultingBody(changes, asset.ConceptPath, repo)
		if err == nil {
			digest := fmt.Sprintf("%x", sha256.Sum256([]byte(repository.NormalizeConceptBody(body))))
			entries := []any{}
			if value, exists := manifest["entries"]; exists {
				var ok bool
				entries, ok = value.([]any)
				if !ok {
					return nil, errors.New("manifest entries must be a list")
				}
			}
			for _, entry := range entries {
				item, ok := entry.(map[string]any)
				if !ok {
					continue
				}
				path, ok := item["path"].(string)
				if ok && item["sha256"] == digest {
					matching = append(matching, path)
				}
			}
			sort.Strings(matching)
			bodyMatches = len(matching) > 0
		} else if !ignorableBundleError(err) {
			return nil, err
		}
		assets = append(assets, map[string]any{"asset_id": asset.AssetID, "concept_path": asset.ConceptPath, "asset_kind": asset.AssetKind, "version": asset.Version, "file_count": manifest["file_count"], "total_uncompressed_bytes": manifest["total_uncompressed_bytes"], "zip_sha256": asset.SHA256, "concept_body_matching_entries": matching, "concept_body_matches_asset": bodyMatches})
	}
	revision, err := repo.main(q.Paths)
	if err != nil {
		return nil, err
	}
	conflicts := []ProposalConflict{}
	if includeConflicts {
		limit := maxProposalSummaryChanges
		conflicts, err = q.conflicts(ctx, record, "", &limit, nil, repo)
		if err != nil {
			return nil, err
		}
	}
	intent := []rune(record.Intent)
	return map[string]any{
		"proposal_id": record.ProposalID, "author_principal": record.AuthorPrincipal, "intent": string(intent[:min(len(intent), maxProposalSummaryText)]), "intent_truncated": len(intent) > maxProposalSummaryText, "status": string(record.Status), "base_revision": record.BaseRevision, "current_revision": revision, "reviewed_by": nullableText(record.ReviewedBy), "applied_operation_id": nullableText(record.AppliedOperationID), "applied_revision": nullableText(record.AppliedRevision), "created_at": record.CreatedAt, "updated_at": record.UpdatedAt, "expires_at": nullableText(record.ExpiresAt), "change_count": len(changes), "changes_truncated": len(changes) > len(summaries), "changes": summaries, "conflicts_included": includeConflicts, "conflicts": conflictDetails(conflicts), "asset_count": len(allAssets), "assets_truncated": len(allAssets) > len(assets), "assets": assets,
	}, nil
}

func (q ProposalQueue) resultingBody(changes []any, path string, repo proposalRepository) (string, error) {
	for _, value := range changes {
		raw, ok := value.(map[string]any)
		if !ok {
			return "", &ChangeValidationError{"proposal change must be an object"}
		}
		if raw["path"] == path && (raw["kind"] == "create" || raw["kind"] == "patch") {
			if body, ok := raw["body"].(string); ok {
				return body, nil
			}
		}
	}
	entry, err := repo.read(q.Paths.CurrentDir, path)
	if err != nil {
		return "", err
	}
	return entry.Document.Body, nil
}
func ignorableBundleError(err error) bool {
	var bundle *repository.BundleError
	var frontmatter *repository.FrontmatterError
	var safety *repository.PathSafetyError
	return errors.Is(err, fs.ErrNotExist) || errors.As(err, &bundle) || errors.As(err, &frontmatter) || errors.As(err, &safety)
}
func conflictDetails(conflicts []ProposalConflict) []any {
	details := []any{}
	for _, c := range conflicts {
		item := map[string]any{"index": c.Index, "status": c.Status, "conflicting_paths": c.ConflictingPaths}
		if c.Reason != "" {
			item["reason"] = c.Reason
		}
		details = append(details, item)
	}
	return details
}

File: go/oracle/proposal_summary.py
"""Capture bounded summaries and concept-body/asset matching without previews."""

from __future__ import annotations

import hashlib
import importlib
from types import SimpleNamespace
from typing import Any
from unittest.mock import patch


def fixtures() -> list[dict[str, Any]]:
    service: Any = importlib.import_module("memento.service")
    proposals: Any = importlib.import_module("memento.control.proposals")
    frontmatter: Any = importlib.import_module("memento.repository.frontmatter")
    cases = []
    for scenario in [
        "empty",
        "rename",
        "missing-fields",
        "nonobject",
        "null-changes",
        "first-body",
        "fallback-body",
        "missing-body",
        "bad-body",
        "no-match",
        "truncated",
        "missing-base",
    ]:
        changes: list[Any] = []
        assets: list[Any] = []
        if scenario == "rename":
            changes = [{"kind": "rename", "path": "/a.md", "new_path": "/b.md"}]
        elif scenario == "missing-fields":
            changes = [{}]
        elif scenario == "nonobject":
            changes = [None, {"kind": "patch", "path": "/a.md"}]
        elif scenario not in {"empty", "null-changes"}:
            changes = [
                {"kind": "patch", "path": "/a.md", "body": "first \r\n"},
                {"kind": "create", "path": "/a.md", "body": "second"},
            ]
            if scenario != "first-body":
                changes = [{"kind": "patch", "path": "/a.md", "body": None}]
            digest = hashlib.sha256(
                b"first" if scenario == "first-body" else b"current"
            ).hexdigest()
            manifest = {
                "file_count": 7,
                "total_uncompressed_bytes": 32,
                "entries": [
                    None,
                    {"path": "z.md", "sha256": digest},
                    {"path": "a.md", "sha256": digest},
                    {"path": "a.md", "sha256": digest},
                    {"path": 1, "sha256": digest},
                    {"path": "other", "sha256": "mismatch"},
                    {},
                ],
            }
            if scenario == "no-match":
                manifest = {}
            assets = [
                SimpleNamespace(
                    asset_id="asset",
                    concept_path="/a.md",
                    asset_kind="docs",
                    version="1.0.0",
                    sha256="zip-sha",
                    manifest=manifest,
                )
            ]
        if scenario == "truncated":
            changes = [{"kind": "patch", "path": f"/{i}.md"} for i in range(201)]
            assets = [
                SimpleNamespace(**{**vars(assets[0]), "asset_id": f"asset-{i}"}) for i in range(201)
            ]
        record = SimpleNamespace(
            proposal_id="proposal",
            author_principal="author",
            intent="😀" * (2001 if scenario == "truncated" else 2000),
            status=proposals.ProposalStatus.APPROVED,
            base_revision="base",
            reviewed_by="reviewer",
            applied_operation_id=None,
            applied_revision=None,
            created_at="created",
            updated_at="updated",
            expires_at=None,
            patch={"changes": None if scenario == "null-changes" else changes},
        )
        runtime = object.__new__(service.MemoryService)
        runtime._deps = SimpleNamespace(
            control_connection=None, repo_paths=SimpleNamespace(current_dir="unused")
        )

        def read(*args: Any, scenario: str = scenario) -> Any:
            if scenario == "missing-body":
                raise FileNotFoundError("missing")
            if scenario == "bad-body":
                raise frontmatter.FrontmatterError("invalid")
            return SimpleNamespace(
                document=SimpleNamespace(
                    body="current \n", frontmatter=SimpleNamespace(id="12345678")
                )
            )

        def diff(*args: Any, scenario: str = scenario, **kwargs: Any) -> tuple[str, ...]:
            if scenario == "missing-base":
                raise service.GitError("missing base")
            return ("/a.md",)

        with (
            patch.object(service, "get_main_revision", lambda _: "main"),
            patch.object(service, "diff_main_paths", diff),
            patch.object(service, "read_bundle_entry", read),
            patch.object(
                service, "list_proposal_assets", lambda *args, assets=assets, **kwargs: assets
            ),
        ):
            for include in [False, True]:
                item: dict[str, Any] = {
                    "scenario": scenario,
                    "include_conflicts": include,
                    "record": {**vars(record), "status": record.status.value},
                    "assets": [vars(asset) for asset in assets],
                }
                try:
                    item["expected"] = runtime._proposal_summary_payload(
                        record, include_conflicts=include
                    )
                except Exception as exc:
                    item["error_type"] = type(exc).__name__
                cases.append(item)
    return cases
